File: shared/go/bridge_client.go
// Package stitch provides shared primitives for Stitch Go clients.
//
// All Go bridge clients (go-python, go-ruby, go-nodejs) import this package
// instead of duplicating scanner, pending-map, kill, and ready-wait logic.
package stitch

import (
	"bufio"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"os/exec"
	"sync"
	"syscall"
	"time"
)

// ─────────────────────────────────────────────────────────────────────────────
// Protocol types
// ─────────────────────────────────────────────────────────────────────────────

// RpcError carries the error payload from the sidecar.
type RpcError struct {
	Message   string `json:"message"`
	Traceback string `json:"traceback,omitempty"`
	Backtrace string `json:"backtrace,omitempty"`
	Code      int    `json:"code,omitempty"`
}

func (e *RpcError) Error() string {
	if e.Traceback != "" {
		return fmt.Sprintf("%s\n%s", e.Message, e.Traceback)
	}
	if e.Backtrace != "" {
		return fmt.Sprintf("%s\n%s", e.Message, e.Backtrace)
	}
	return e.Message
}

// RpcResponse is the wire format for incoming JSON-RPC responses.
type RpcResponse struct {
	ID     string          `json:"id"`
	Result json.RawMessage `json:"result,omitempty"`
	Error  *RpcError       `json:"error,omitempty"`
}

// ─────────────────────────────────────────────────────────────────────────────
// PendingMap - mutex-guarded map of in-flight calls
// ─────────────────────────────────────────────────────────────────────────────

// PendingMap manages in-flight RPC calls, each identified by a string UUID.
// It is safe for concurrent use from multiple goroutines.
type PendingMap struct {
	mu    sync.Mutex
	calls map[string]chan RpcResponse
}

// NewPendingMap creates an initialised PendingMap.
func NewPendingMap() *PendingMap {
	return &PendingMap{calls: make(map[string]chan RpcResponse)}
}

// Register inserts a channel for the given request id and returns it.
// The returned channel is buffered with capacity 1.
func (p *PendingMap) Register(id string) chan RpcResponse {
	ch := make(chan RpcResponse, 1)
	p.mu.Lock()
	p.calls[id] = ch
	p.mu.Unlock()
	return ch
}

// Dispatch delivers resp to the channel registered for resp.ID.
// It is a no-op if no channel is registered (e.g. caller timed out).
func (p *PendingMap) Dispatch(resp RpcResponse) {
	p.mu.Lock()
	ch, ok := p.calls[resp.ID]
	if ok {
		delete(p.calls, resp.ID)
	}
	p.mu.Unlock()
	if ok {
		ch <- resp
	}
}

// Delete removes the entry for id without delivering a response.
// Use this when a call is cancelled or timed out before a reply arrives.
func (p *PendingMap) Delete(id string) {
	p.mu.Lock()
	delete(p.calls, id)
	p.mu.Unlock()
}

// DrainWithError unblocks all pending callers with an error response.
// Call this when the child process exits unexpectedly.
func (p *PendingMap) DrainWithError(message string) {
	p.mu.Lock()
	for id, ch := range p.calls {
		ch <- RpcResponse{ID: id, Error: &RpcError{Message: message}}
		delete(p.calls, id)
	}
	p.mu.Unlock()
}

// ─────────────────────────────────────────────────────────────────────────────
// Scanner constructor with enlarged buffer
// ─────────────────────────────────────────────────────────────────────────────

const defaultScannerBufSize = 4 * 1024 * 1024 // 4 MiB

// NewScanner returns a *bufio.Scanner reading from r with a 4 MiB buffer.
// This avoids scanner.ErrTooLong on large JSON payloads.
func NewScanner(r io.Reader) *bufio.Scanner {
	s := bufio.NewScanner(r)
	s.Buffer(make([]byte, defaultScannerBufSize), defaultScannerBufSize)
	return s
}

// ─────────────────────────────────────────────────────────────────────────────
// KillChild - SIGTERM → SIGKILL(2 s) pattern
// ─────────────────────────────────────────────────────────────────────────────

// KillChild sends SIGTERM to the child process; if it has not exited after 2 s
// it sends SIGKILL.  Safe to call when cmd.Process is nil.
func KillChild(cmd *exec.Cmd) {
	if cmd.Process == nil {
		return
	}

	_ = cmd.Process.Signal(syscall.SIGTERM)

	done := make(chan struct{})
	go func() {
		_ = cmd.Wait()
		close(done)
	}()

	timer := time.NewTimer(2 * time.Second)
	defer timer.Stop()
	select {
	case <-done:
	case <-timer.C:
		_ = cmd.Process.Kill()
		<-done
	}
}

// ─────────────────────────────────────────────────────────────────────────────
// WaitReady - read lines until {"ready":true}
// ─────────────────────────────────────────────────────────────────────────────

// WaitReady advances scanner until it finds a line that decodes as
// {"ready":true}.  Returns an error if the scanner ends before that.
func WaitReady(scanner *bufio.Scanner) error {
	for scanner.Scan() {
		var msg map[string]interface{}
		if err := json.Unmarshal(scanner.Bytes(), &msg); err != nil {
			continue // not valid JSON, keep reading
		}
		if ready, _ := msg["ready"].(bool); ready {
			return nil
		}
	}
	if err := scanner.Err(); err != nil {
		return fmt.Errorf("stitch: waiting for ready: %w", err)
	}
	return errors.New("stitch: child closed stdout before sending ready signal")
}

File: shared/rust/bridge_client.rs
//! Stitch - shared Rust bridge client module.
//!
//! All Rust bridge clients (rust-python, rust-go, rust-ruby) include this
//! module (`mod bridge_client; use bridge_client::*;`) rather than duplicating
//! the pending-map, reader-thread spawn, kill helper, and error types.

use std::{
    collections::HashMap,
    io::{BufRead, BufReader},
    process::{Child, ChildStdout},
    sync::{
        mpsc::{sync_channel, SyncSender},
        Arc, Mutex,
    },
    thread::{self, JoinHandle},
};

use serde::{Deserialize, Serialize};
use serde_json::Value;

// ─────────────────────────────────────────────────────────────────────────────
// Wire types
// ─────────────────────────────────────────────────────────────────────────────

/// JSON-RPC error object carried in an error response from the sidecar.
#[derive(Debug, Deserialize, Clone)]
pub struct RpcError {
    pub code: Option<i64>,
    pub message: String,
    pub traceback: Option<String>,
    pub backtrace: Option<String>,
}

impl std::fmt::Display for RpcError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.message)
    }
}

/// A parsed response line from the sidecar (success or error, identified by id).
#[derive(Debug, Clone)]
pub struct RpcResponse {
    pub id: String,
    pub result: Option<Value>,
    pub error: Option<RpcError>,
}

// ─────────────────────────────────────────────────────────────────────────────
// PendingMap
// ─────────────────────────────────────────────────────────────────────────────

/// Shared pending-call map.  Both the caller thread and the reader thread hold
/// a clone of this `Arc` - the caller inserts, the reader removes and delivers.
pub type PendingMap = Arc<Mutex<HashMap<String, SyncSender<RpcResponse>>>>;

/// Create an empty PendingMap.
pub fn new_pending_map() -> PendingMap {
    Arc::new(Mutex::new(HashMap::new()))
}

// ─────────────────────────────────────────────────────────────────────────────
// Reader thread
// ─────────────────────────────────────────────────────────────────────────────

/// Spawn a daemon thread that reads newline-delimited JSON from `stdout`,
/// dispatches each response to the matching `SyncSender` in `pending`, and
/// signals `ready_tx` when the first `{"ready":true}` line is seen.
///
/// The returned `JoinHandle` should be stored in the bridge struct so the
/// thread is joined (or at least kept alive) for the bridge's lifetime.
pub fn spawn_reader_thread(
    stdout: ChildStdout,
    pending: PendingMap,
    ready_tx: SyncSender<()>,
) -> JoinHandle<()> {
    thread::spawn(move || {
        let reader = BufReader::new(stdout);
        let mut ready_sent = false;

        for line in reader.lines() {
            let raw = match line {
                Ok(l) => l,
                Err(e) => {
                    eprintln!("[bridge_client] reader IO error: {e}");
                    break;
                }
            };

            let trimmed = raw.trim();
            if trimmed.is_empty() {
                continue;
            }

            let v: Value = match serde_json::from_str(trimmed) {
                Ok(v) => v,
                Err(e) => {
                    eprintln!("[bridge_client] malformed JSON: {e} - `{trimmed}`");
                    continue;
                }
            };

            // Ready signal
            if !ready_sent && v.get("ready") == Some(&Value::Bool(true)) {
                ready_sent = true;
                let _ = ready_tx.send(());
                continue;
            }

            // Normal RPC response
            let id = match v.get("id").and_then(Value::as_str) {
                Some(s) => s.to_string(),
                None => {
                    eprintln!("[bridge_client] response missing `id`: {v}");
                    continue;
                }
            };

            let error = v.get("error").and_then(|e| {
                serde_json::from_value::<RpcError>(e.clone()).ok()
            });
            let result = v.get("result").cloned();
            let resp = RpcResponse { id: id.clone(), result, error };

            let mut map = pending.lock().unwrap();
            if let Some(tx) = map.remove(&id) {
                let _ = tx.send(resp);
            } else {
                eprintln!("[bridge_client] unknown response id: {id}");
            }
        }

        // Stdout closed - drain pending callers with an error.
        if !ready_sent {
            let _ = ready_tx.send(());
        }
        let mut map = pending.lock().unwrap();
        for (id, tx) in map.drain() {
            let _ = tx.send(RpcResponse {
                id,
                result: None,
                error: Some(RpcError {
                    code: Some(-32000),
                    message: "child process exited unexpectedly".to_string(),
                    traceback: None,
                    backtrace: None,
                }),
            });
        }
    })
}

// ─────────────────────────────────────────────────────────────────────────────
// kill_child helper
// ─────────────────────────────────────────────────────────────────────────────

/// Kill the child process and wait for it to exit.
/// On Unix: sends SIGKILL directly (callers should close stdin first for a
/// graceful shutdown before resorting to kill_child).
pub fn kill_child(child: &mut Child) {
    let _ = child.kill();
    let _ = child.wait();
}

// ─────────────────────────────────────────────────────────────────────────────
// Convenience: make a pending-slot and SyncSender pair
// ─────────────────────────────────────────────────────────────────────────────

/// Register a new call id in `pending` and return the receiver end of a
/// one-shot channel.  The reader thread will deliver to the sender end.
pub fn register_call(
    pending: &PendingMap,
    id: &str,
) -> std::sync::mpsc::Receiver<RpcResponse> {
    let (tx, rx) = sync_channel::<RpcResponse>(1);
    pending.lock().unwrap().insert(id.to_string(), tx);
    rx
}
